mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,357 @@
+        Self {
+            inner: orders
+                .iter()
+                .map(|order| {
+                    ((src_token.clone(), order.coin.clone()), LrData {
+                        order: order.clone(),
+                        src_contract: None,
+                        src_decimals: None,
+                        src_amount: None,
+                        dst_contract: None,
+                        dst_amount: None,
+                        dst_decimals: None,
+                        chain_id: None,
+                        lr_price: None,
+                        lr_swap_data: None,
+                    })
+                })
+                .collect::<HashMap<(_, _), _>>(),
+        }
+    }
+
+    /// Calculate amounts of destination tokens required to fill ask orders for the requested base_amount:
+    /// multiplies base_amount by the order price (base_amount must be in 'coins')
+    async fn calc_destination_token_amounts(
+        &mut self,
+        ctx: &MmArc,
+        base_amount: &MmNumber,
+    ) -> MmResult<(), ApiIntegrationRpcError> {
+        for lr_data in self.inner.values_mut() {
+            let price: MmNumber = lr_data.order.price.rational.clone().into();
+            let dst_amount = base_amount * &price;
+            let coin = lp_coinfind_or_err(ctx, &lr_data.order.coin).await?;
+            lr_data.dst_amount = Some(wei_from_coins_mm_number(&dst_amount, coin.decimals())?);
+            log::debug!(
+                "calc_destination_token_amounts lr_data.order.coin={} coin.decimals()={} lr_data.dst_amount={:?}",
+                lr_data.order.coin,
+                coin.decimals(),
+                lr_data.dst_amount
+            );
+        }
+        Ok(())
+    }
+
+    fn update_with_lr_prices(&mut self, lr_prices: &HashMap<(Ticker, Ticker), MmNumber>) {
+        for (key, val) in self.inner.iter_mut() {
+            val.lr_price = lr_prices.get(key).cloned();
+        }
+    }
+
+    fn update_with_lr_swap_data(&mut self, lr_swap_data: &HashMap<(Ticker, Ticker), ClassicSwapData>) {
+        for (key, val) in self.inner.iter_mut() {
+            val.lr_swap_data = lr_swap_data.get(key).cloned();
+        }
+    }
+
+    async fn update_with_contracts(&mut self, ctx: &MmArc) -> MmResult<(), ApiIntegrationRpcError> {
+        for ((src_token, dst_token), lr_data) in self.inner.iter_mut() {
+            let (src_coin, src_contract) = get_coin_for_one_inch(ctx, src_token).await?;
+            let (dst_coin, dst_contract) = get_coin_for_one_inch(ctx, dst_token).await?;
+            lr_data.src_contract = EthAddress::from_str(&src_contract).ok(); // TODO: return error?
+            lr_data.dst_contract = EthAddress::from_str(&dst_contract).ok();
+            lr_data.src_decimals = Some(src_coin.decimals()); // TODO: maybe we should check for 0 and return error (if missed in coins)? Now it works okay if decimals omitted in coins
```

Comment:
Can we check this some where earlier? during coin activation or smth else

Reply:
We may check this earlier too but maybe good to have all checks (chain_id, decimals, contract) in one place where we use these data.

Reply:
decimals check added here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2362/commits/32ddfa21b0656e4881b43d2bc59ee5e6516bd9de
note resolved